File: libs/language/src/lib/merge/processors/BasicTypesVerifier.ts
import { BlueNode } from '../../model';
import { NodeProvider } from '../../NodeProvider';
import { MergingProcessor } from '../MergingProcessor';
import { NodeTypes } from '../../utils';

/**
 * Verifies that nodes with basic types don't have items or properties
 */
export class BasicTypesVerifier implements MergingProcessor {
  process(target: BlueNode): BlueNode {
    return target;
  }

  postProcess(
    target: BlueNode,
    source: BlueNode,
    nodeProvider: NodeProvider
  ): BlueNode {
    const targetType = target.getType();
    if (
      targetType !== undefined &&
      NodeTypes.isSubtypeOfBasicType(targetType, nodeProvider)
    ) {
      const items = target.getItems();
      const properties = target.getProperties();

      if (
        (items !== undefined && items.length > 0) ||
        (properties !== undefined && Object.keys(properties).length > 0)
      ) {
        const basicTypeName = NodeTypes.findBasicTypeName(
          targetType,
          nodeProvider
        );
        const typeName = targetType.getName() || 'unknown';
        throw new Error(
          `Node of type "${typeName}" (which extends basic type "${basicTypeName}") must not have items, properties or contracts.`
        );
      }
    }
    return target;
  }
}

File: references/java-sdk/src/main/java/blue/language/merge/processor/BasicTypesVerifier.java
package blue.language.merge.processor;

import blue.language.merge.MergingProcessor;
import blue.language.NodeProvider;
import blue.language.merge.NodeResolver;
import blue.language.model.Node;
import blue.language.utils.Types;

import static blue.language.utils.Types.findBasicTypeName;

public class BasicTypesVerifier implements MergingProcessor {
    @Override
    public void process(Node target, Node source, NodeProvider nodeProvider, NodeResolver nodeResolver) {
        // do nothing
    }

    @Override
    public void postProcess(Node target, Node source, NodeProvider nodeProvider, NodeResolver nodeResolver) {
        if (target.getType() != null && Types.isSubtypeOfBasicType(target.getType(), nodeProvider)) {
            if ((target.getItems() != null && !target.getItems().isEmpty()) ||
                (target.getProperties() != null && !target.getProperties().isEmpty())) {
                String basicTypeName = findBasicTypeName(target.getType(), nodeProvider);
                throw new IllegalArgumentException("Node of type \"" + target.getType().getName() +
                                                   "\" (which extends basic type \"" + basicTypeName +
                                                   "\") must not have items or properties.");
            }
        }
    }
}
